### PR TITLE
[WIP][FIX] Fix Mixtral model

### PR DESCRIPTION
**Pull request by casper-hansen**
This PR is a WIP based on @kalomaze's implementation that fixes the Mixtral model. It has been known for a while that Mixtral has been hard to train due to some bug in the code. Please note this is meant for testing based on the provided code.

More context:
- https://twitter.com/kalomaze/status/1786869036946522256
- https://gist.github.com/kalomaze/661b79095fdd91df8a84802f7cb6f26a

### CODE REVIEW

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # coding=utf-8
-# Copyright 2023 Mistral AI and the HuggingFace Inc. team. All rights reserved.
+# Copyright 2023 Mixtral AI and the HuggingFace Inc. team. All rights reserved.
```
Comment (PhilipMay) [inline src/transformers/models/mixtral/modeling_mixtral.py:2]: The name of the company is Mistral AI and not Mixtral AI.
Also see: https://mistral.ai/company/

This line should not be changed IMO.

**Comment by PhilipMay** [comment]
See discussion with @kalomaze here on DIscors: https://discord.com/channels/1053877538025386074/1132352574750728192/1236763822081966100

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 1 above)
Comment (kalomaze) [reply src/transformers/models/mixtral/modeling_mixtral.py:2]: Some of the documentation was CTRL+H replaced by accident when I was making my changes and should be restored, yeah

**Comment by kalomaze** [comment]
> This PR is a WIP based on @kalomaze's implementation that fixes the Mixtral model. It has been known for a while that Mixtral has been hard to train due to some bug in the code. Please note this is meant for testing based on the provided code.
> 
> More context:
> 
> * https://twitter.com/kalomaze/status/1786869036946522256
> * https://gist.github.com/kalomaze/661b79095fdd91df8a84802f7cb6f26a

Thanks for showing interest and putting up a PR.

What I specifically did to the code:
- Replaced the Attention, RoPE, etc functions (that _should_ be functionally equivalent to Llama architecture) with their equivalents from modeling_llama.py.
- MLP evaluation / MoE related components, load balancing, etc are left unchanged, as they were not the cause of high loss.

Then I confirmed training was working as intended by testing topk=1 full finetuning with frozen routing on a "fake MoE" made with mergekit (in which all MLPs are identical).

image

I'm told what specifically fixed the training on my end was replacing the RoPE / Rotary embeddings functions with the Llama implementation equivalents (according to @2wlearning on Twitter):
image

If there is a need to _only_ keep the necessary changes rather than fully standardizing the implementation of Non-MoE components (as they are seen in modeling_llama.py), the Transformers maintainers may want to look into adopting the RoPE related fixes.

From my perspective though, the architectures are essentially equivalent outside of the MoE components; not to mention, the SWA component has been abandoned by Mistral for Mixtral 8x7b/8x22b/7b-v0.2, in favor of higher native context lengths. So I don't see a downside to standardizing these functionally equivalent functions.

Opinions @ArthurZucker ?

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 2]
```diff
@@ -323,57 +367,54 @@ def forward(
...
+            query_states = [F.linear(hidden_states, query_slices[i]) for i in range(self.config.pretraining_tp)]
+            query_states = torch.cat(query_states, dim=-1)
+
+            key_states = [F.linear(hidden_states, key_slices[i]) for i in range(self.config.pretraining_tp)]
+            key_states = torch.cat(key_states, dim=-1)
+
+            value_states = [F.linear(hidden_states, value_slices[i]) for i in range(self.config.pretraining_tp)]
+            value_states = torch.cat(value_states, dim=-1)
+
+        else:
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:392]: pretraining TP should not be added no? No one uses this and its slow

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 3]
```diff
@@ -387,25 +428,29 @@ def forward(
...
         attn_output = attn_output.transpose(1, 2).contiguous()
+
         attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)
 
-        attn_output = self.o_proj(attn_output)
+        if self.config.pretraining_tp > 1:
+            attn_output = attn_output.split(self.hidden_size // self.config.pretraining_tp, dim=2)
+            o_proj_slices = self.o_proj.weight.split(self.hidden_size // self.config.pretraining_tp, dim=1)
+            attn_output = sum([F.linear(attn_output[i], o_proj_slices[i]) for i in range(self.config.pretraining_tp)])
+        else:
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:438]: same comment here, I don't think this is what allowed for better training right?

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 4]
```diff
@@ -868,11 +811,18 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
             expert_layer = self.experts[expert_idx]
             idx, top_x = torch.where(expert_mask[expert_idx])
 
+            if top_x.shape[0] == 0:
+                continue
+
+            # in torch it is faster to index using lists than torch tensors
+            top_x_list = top_x.tolist()
+            idx_list = idx.tolist()
+
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:820]: 😢 did this also help? We removed it to support #29858. Would also be supprised that this is causing issues. Could you detail what is going on here?

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 5]
```diff
@@ -1002,6 +955,27 @@ def _init_weights(self, module):
...
+                dtype = self.config._pre_quantization_dtype
+            else:
+                dtype = layer.self_attn.o_proj.weight.dtype
+            layer.self_attn.past_key_value = cache_cls(
+                self.config, max_batch_size, max_cache_len, device=device, dtype=dtype
+            )
+
+    def _reset_cache(self):
+        for layer in self.model.layers:
+            layer.self_attn.past_key_value = None
```
Comment (ArthurZucker) [inline src/transformers/models/mixtral/modeling_mixtral.py:977]: cache was moved outside the model in #30476

**Review by ArthurZucker — COMMENTED** [review]
Thanks, could you isolate the changes (I believe it's mostly propagating the ROPE changes from Gemma and Llama to mistral) that improved the training? 

Also a similar PR is done in #30642 to support torch compile, we can disentangle both imo

**Comment by casper-hansen** [comment]
> Thanks, could you isolate the changes (I believe it's mostly propagating the ROPE changes from Gemma and Llama to mistral) that improved the training?
> 
> Also a similar PR is done in #30642 to support torch compile, we can disentangle both imo

Yes, the isolated changes should be RoPE based on the comments left here. I think the next step will be to start over with the modeling here and copy in the RoPE code from Llama.

**Comment by claralp** [comment]
> What I specifically did to the code:
> 
> * Replaced the Attention, RoPE, etc functions (that _should_ be functionally equivalent to Llama architecture) with their equivalents from modeling_llama.py.
> * MLP evaluation / MoE related components, load balancing, etc are left unchanged, as they were not the cause of high loss.
> 
> Then I confirmed training was working as intended by testing topk=1 full finetuning with frozen routing on a "fake MoE" made with mergekit (in which all MLPs are identical).
> 
> image
> 


Can you show the model config you used for these experiments? e.g. did you used RoPE scaling here

**Comment by kalomaze** [comment]
> there is a problem with the new code (loss goes up and then does not move anymore). Can you please clarify?

I think you are extrapolating a bit too much here. It is not instruction finetuning data, and instead general webscrape data from FineWeb, which will generally take a long time and large batches for loss to progress further as it is not a typical finetune.
I also used a small batch size of 2.

Notice how the old code seemed to plateau; in my other test, it wouldn't go below 2.3 after ~100 million tokens.

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 2 above)
Comment (kalomaze) [reply src/transformers/models/mixtral/modeling_mixtral.py:392]: I swapped out several functions at a time to try to narrow it down so that's why there might be irrelevant changes
as such, anything outside of RoPE probably doesn't need* to be adjusted, as that was what was causing divergence [according to what @2wlearning told me on Twitter]

**Comment by kalomaze** [comment]
> e.g. did you used RoPE scaling here

I used the RoPE config as specified in Llama3 8b's config.json:
```
  "rope_scaling": null,
  "rope_theta": 500000.0,
```

In 8x7b config.json, you can find

```
  "rope_theta": 1000000.0,
```

And there is no mention of a `"rope_scaling"` type, so I presumed that it similarly registers as null in actual use.

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 4 above)
Comment (kalomaze) [reply src/transformers/models/mixtral/modeling_mixtral.py:820]: I don't think it's this, as I mentioned, most probably related to the rotary embeddings related code which was inherited all the way back [from GPT-Neo](https://twitter.com/rafacelente/status/1787229662533357922)

**Comment by casper-hansen** [comment]
I have now updated the code to focus mostly on RoPE and based it on the current Mixtral model on the main branch. @kalomaze can you confirm if this the intended fix?

**Comment by casper-hansen** [comment]
I have validated that perplexity is lower than main branch using the gist that was posted (with a small fix). However, when trying to extract the most essential changes to RoPE scaling, I must have taken out the part that lowered the perplexity score. I am looking to create a working commit. @kalomaze @DreamGenX 

Perplexity:
- Main: 3.843
- PR: 2.619

I will be back with a commit where this can be fully verified!

**Comment by casper-hansen** [comment]
It seems there is no measureable difference on eager/flash attention. @kalomaze This suggests you trained your model using SDPA, and the reason you got a better loss was because `is_causal` was set to False.

Perplexity (PR):
- `attn_implementation="sdpa"`: 2.619
- `attn_implementation="eager"`: 3.843
- `attn_implementation="flash_attention_2"`: 3.843

Now, what makes SDPA have a much lower perplexity? I investigated and found that `is_causal` on SDPA is set to False by default, which causes the low perplexity. However, on the main branch, it's determined by the following. @ArthurZucker please advise if you have any idea about what is happening here.

https://github.com/huggingface/transformers/blob/75bbfd5b2237b7e35a9265731ecf63022579e7e2/src/transformers/models/mixtral/modeling_mixtral.py#L765-L773

<details>

<summary>ppl.py script</summary>

```python
import torch
import torch.nn as nn
from tqdm import tqdm
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer


def evaluate_perplexity(model, tokenizer):
    def _perplexity(nlls, n_samples, seqlen):
        return torch.exp(torch.stack(nlls).sum() / (n_samples * seqlen))

    # load and prepare dataset
    data = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
    data = tokenizer("\n\n".join(data["text"]), return_tensors="pt")
    data = data.input_ids.to(model.device)

    seqlen = 2048
    model = model.eval()
    n_samples = data.numel() // seqlen

    nlls = []

    with tqdm(range(n_samples), desc="Perplexity -") as progress_bar:
        for i in progress_bar:
            start_index = i * seqlen
            end_index = (i + 1) * seqlen
            batch = data[:, start_index:end_index].to(model.device)
            with torch.no_grad():
                logits = model(batch).logits
            shift_logits = logits[:, :-1, :].contiguous().float()
            shift_labels = data[:, start_index:end_index][:, 1:]
            loss_fct = nn.CrossEntropyLoss()
            loss = loss_fct(
                shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
            )
            neg_log_likelihood = loss.float() * seqlen
            nlls.append(neg_log_likelihood)

            curr_ppl = _perplexity(nlls, i + 1, seqlen)
            progress_bar.set_description(f"Perplexity {curr_ppl:.3f}")

    ppl = _perplexity(nlls, n_samples, seqlen)

    return ppl.item()

if __name__ == "__main__":
    model_path = 'mistralai/Mixtral-8x7B-v0.1'
    model = AutoModelForCausalLM.from_pretrained(
        model_path,
        device_map="auto",
        torch_dtype=torch.bfloat16,
        attn_implementation="flash_attention_2",
    )
    tokenizer = AutoTokenizer.from_pretrained(model_path)
    evaluate_perplexity(model, tokenizer)
```

</details>

**Comment by casper-hansen** [comment]
Closing this PR for now as it seems to be a bust. `is_causal` is to make sure that SDPA correctly masks the inputs, so setting it to False will obviously give lower perplexity.

**Comment by DreamGenX** [comment]
@casper-hansen Thank you so much for your dillegent investigation

**Comment by ArthurZucker** [comment]
The idea is to dispatch to the memory efficient attention kernel of sdpa. When the attention mask is None, then you are using causal attention by default if you are in the prefill phase. 
However, if you are decoding, and the attention_mask is stil None (no padding) then the attention becomes non causal: you pay attention to everything that happened before -> thus if q_len == 1 and attention_mask is None you should not be causal. 

Results should however be the same. I am not at the origin of this commit, cc @fxmarty as this sounds interesting to investigate.
